### PR TITLE
Live test pipelines

### PR DESCRIPTION
**Pull request by danieljurek**
(had to create live tests pipelines in a branch from the main repo to validate pipeline creation scenarios) 

@weshaggard, @mitchdenny -- This is a simple example of how we can integrate live testing with Unified Pipelines. 

I've created a `- tests` pipeline and modified the unified pipeline to support this. 
* `tests` -- Uses a variable called `PipelineType` with value `tests` set as a variable on the pipeline which cannot be overridden.. This supports running live tests from PR comments (`/azp run c - client - tests`). This skips unified pipeline flow (CI build, release stages) ([example execution](https://dev.azure.com/azure-sdk/internal/_build/results?buildId=406743&view=results))
* Unified pipeline -- Added the `PipelineType` variable here with the value `unified`. This skips live tests and allows for releases. ([example execution](https://dev.azure.com/azure-sdk/internal/_build/results?buildId=406748&view=results))

Canceling builds took a little while to figure out. If I'm understanding canceling correctly one must cancel individual stages as opposed to canceling the build all-up.

### CODE REVIEW

File: eng/pipelines/templates/stages/archetype-sdk-client.yml [hunk 1]
```diff
@@ -1,10 +1,25 @@
...
+      dependsOn: []
+      jobs:
+        - template: /eng/pipelines/templates/jobs/archetype-sdk-tests.yml
+          parameters:
+            LiveTestSamples:
+              - Name: Storage Blobs
+                LinuxMacPath: ./build/sdk/storage/blobs/samples/blobs_client_example
+                WindowsPath: ./build/sdk/storage/blobs/samples/debug/blobs_client_example.exe
 
+  - ${{ if and(ne(variables['Build.Reason'], 'PullRequest'), eq(variables['System.TeamProject'], 'internal')) }}:
```
Comment (weshaggard) [inline eng/pipelines/templates/stages/archetype-sdk-client.yml:22]: This should be Build.Reason = Manual and Internal

File: eng/pipelines/templates/stages/archetype-sdk-client.yml [hunk 2]
```diff
@@ -1,10 +1,25 @@
 stages:
   - stage: Build
```
Comment (weshaggard) [inline eng/pipelines/templates/stages/archetype-sdk-client.yml:2]: Add template condition around and exclude if pipeline name ends with "- tests"

File: eng/pipelines/templates/stages/archetype-sdk-client.yml [hunk 3]
```diff
@@ -1,10 +1,25 @@
...
+    - stage: LiveTest
+      # Purpose must be set at the level of the pipeline to enable this scenario
+      condition: eq(variables['PipelineType'], 'tests')
+      dependsOn: []
+      jobs:
+        - template: /eng/pipelines/templates/jobs/archetype-sdk-tests.yml
+          parameters:
+            LiveTestSamples:
+              - Name: Storage Blobs
+                LinuxMacPath: ./build/sdk/storage/blobs/samples/blobs_client_example
```
Comment (danieljurek) [inline eng/pipelines/templates/stages/archetype-sdk-client.yml:19]: Change to NonWindowsPath

File: eng/pipelines/templates/stages/archetype-sdk-client.yml [hunk 4]
```diff
@@ -1,10 +1,25 @@
 stages:
   - stage: Build
+    dependsOn: []
+    # Purpose must be set at the level of the pipeline to enable this scenario
+    condition: eq(variables['PipelineType'], 'unified')
     jobs:
       - template: ../jobs/archetype-sdk-client.yml
 
-  - ${{if and(ne(variables['Build.Reason'], 'PullRequest'), eq(variables['System.TeamProject'], 'internal'))}}:
+  - ${{ if eq(variables['System.TeamProject'], 'internal') }}:
```
Comment (weshaggard) [inline eng/pipelines/templates/stages/archetype-sdk-client.yml:9]: Add pipeline names ends with "- tests"

File: eng/pipelines/templates/stages/archetype-sdk-client.yml [hunk 5]
```diff
@@ -1,10 +1,25 @@
...
+    - stage: LiveTest
+      # Purpose must be set at the level of the pipeline to enable this scenario
+      condition: eq(variables['PipelineType'], 'tests')
```
Comment (weshaggard) [inline eng/pipelines/templates/stages/archetype-sdk-client.yml:12]: Can we add a Skip variable for the condition.

File: eng/pipelines/templates/jobs/archetype-sdk-tests.yml [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+          CMAKE_GENERATOR_PLATFORM: x64
+          build.args: ' -DTRANSPORT_CURL=ON'
+        MacOS_x64_with_samples:
+          vm.image: 'macOS-10.15'
+          vcpkg.deps: 'curl[ssl]'
+          VCPKG_DEFAULT_TRIPLET: 'x64-osx'
+          build.args: ' -DTRANSPORT_CURL=ON'
+
+    pool:
+      vmImage: $(vm.image)
```
Comment (weshaggard) [inline eng/pipelines/templates/jobs/archetype-sdk-tests.yml:34]: This should be OSVMImage

File: eng/pipelines/templates/steps/vcpkg.yml [hunk 7]
```diff
@@ -0,0 +1,47 @@
...
+        if [ $? -ne 0 ]
+        then
+          echo "./bootstrap-vcpkg.sh FAILED"
+          exit 1
+        fi
+
+        echo "##vso[task.prependpath]$(pwd)"
+        echo "##vso[task.setvariable variable=VCPKG_INSTALLATION_ROOT;]$(pwd)"
+    # Execute only for Mac and if there is at least one dependency to be installed
+    condition: and(succeeded(), contains(variables['${{ parameters.VmImageVariableName }}'], 'macOS'), not(eq(variables['${{ parameters.DependenciesVariableName }}'], '')))
```
Comment (weshaggard) [inline eng/pipelines/templates/steps/vcpkg.yml:40]: Just just use OSVMImage as the variable name as that is the common name we use.

File: eng/pipelines/templates/jobs/archetype-sdk-tests.yml [hunk 8]
```diff
@@ -0,0 +1,63 @@
...
+
+      # Run live tests
+      - ${{ each sample in parameters.LiveTestSamples }}:
+        - script: ${{ sample.LinuxMacPath }}
+          displayName: Live Test ${{ sample.Name }} (Non-Windows)
+          condition: not(contains(variables['vm.image'], 'windows'))
+
+        - pwsh: ${{ sample.WindowsPath }}
+          displayName: Live Test ${{ sample.Name }} (Windows)
+          condition: contains(variables['vm.image'], 'windows')
```
Comment (weshaggard) [inline eng/pipelines/templates/jobs/archetype-sdk-tests.yml:58]: OSVMImage

File: eng/pipelines/templates/jobs/archetype-sdk-tests.yml [hunk 9]
```diff
@@ -0,0 +1,63 @@
...
+        parameters:
+          ServiceDirectory: '.'
+
+      # Run live tests
+      - ${{ each sample in parameters.LiveTestSamples }}:
+        - script: ${{ sample.LinuxMacPath }}
+          displayName: Live Test ${{ sample.Name }} (Non-Windows)
+          condition: not(contains(variables['vm.image'], 'windows'))
+
+        - pwsh: ${{ sample.WindowsPath }}
```
Comment (weshaggard) [inline eng/pipelines/templates/jobs/archetype-sdk-tests.yml:56]: Why not just use script on all os's?

File: eng/pipelines/templates/stages/archetype-sdk-client.yml (hunk 4 above)
Comment (danieljurek) [reply eng/pipelines/templates/stages/archetype-sdk-client.yml:9]: If we do this then the only time this pipeline runs is in the context of tests... There will be no point in skipping it (below)

File: eng/pipelines/templates/stages/archetype-sdk-client.yml (hunk 5 above)
Comment (danieljurek) [reply eng/pipelines/templates/stages/archetype-sdk-client.yml:12]: It's possible but the only situation where it'd be used is when running the live tests pipeline if we also include the proposal above to only run live tests in the ` - tests` pipeline.

File: eng/pipelines/templates/stages/archetype-sdk-client.yml (hunk 4 above)
Comment (weshaggard) [reply eng/pipelines/templates/stages/archetype-sdk-client.yml:9]: Correct we don't want the " - tests" condition here (too many iterations in my head).

File: eng/pipelines/templates/jobs/archetype-sdk-client.yml [hunk 10]
```diff
@@ -210,67 +210,24 @@ jobs:
...
-
-
-  - script: |
-      vcpkg install $(vcpkg.deps)
-    displayName: vcpkg install dependencies
-    # Execute only if there is at least one dependency to be installed
-    condition: and(succeeded(), not(eq(variables['vcpkg.deps'], '')))
+  - template: /eng/pipelines/templates/steps/vcpkg.yml
+    parameters:
+      VmImageVariableName: OSVmImage
```
Comment (weshaggard) [inline eng/pipelines/templates/jobs/archetype-sdk-client.yml:215]: I wouldn't expect we need to pass this down any more given we are using a well known variable for the OS.

File: eng/pipelines/templates/stages/archetype-sdk-client.yml [hunk 11]
```diff
@@ -1,10 +1,23 @@
...
-  - ${{if and(ne(variables['Build.Reason'], 'PullRequest'), eq(variables['System.TeamProject'], 'internal'))}}:
+  - ${{ if eq(variables['System.TeamProject'], 'internal') }}:
+    - stage: LiveTest
+      dependsOn: []
+      jobs:
+        - template: /eng/pipelines/templates/jobs/archetype-sdk-tests.yml
+          parameters:
+            LiveTestSamples:
+              - Name: Storage Blobs
+                NonWindowsPath: ./build/sdk/storage/blobs/samples/blobs_client_example
```
Comment (weshaggard) [inline eng/pipelines/templates/stages/archetype-sdk-client.yml:17]: I believe we should be able to use the NonWindowsPath on windows as well.

Comment (danieljurek) [reply eng/pipelines/templates/stages/archetype-sdk-client.yml:17]: NonWindowsPath differs in where the cmake generated projects output the executables. In the case of VS projects (for windows path) the exe is generated in the `Debug` folder.

Comment (danieljurek) [reply eng/pipelines/templates/stages/archetype-sdk-client.yml:17]: We could potentially compose it by having path prefix and binary name parameters but we'll end up with the same number of parameters and it'd be less flexible.

File: eng/pipelines/templates/stages/archetype-sdk-client.yml [hunk 12]
```diff
@@ -1,10 +1,23 @@
...
+                WindowsPath: build\sdk\storage\blobs\samples\debug\blobs_client_example.exe
```
Comment (ahsonkhan) [inline eng/pipelines/templates/stages/archetype-sdk-client.yml:18]: If I want to add more samples to test, do I extend this `LiveTestSamples` list and hard-code the path to the built executable?

File: eng/pipelines/templates/jobs/archetype-sdk-tests.yml [hunk 13]
```diff
@@ -0,0 +1,64 @@
...
+          condition: not(contains(variables['OSVmImage'], 'windows'))
+
+        - script: ${{ sample.WindowsPath }}
+          displayName: Live Test ${{ sample.Name }} (Windows)
+          condition: contains(variables['OSVmImage'], 'windows')
+
+
+      - template: /eng/common/TestResources/remove-test-resources.yml
+        parameters:
+          ServiceDirectory: '.'
```
Comment (ahsonkhan) [inline eng/pipelines/templates/jobs/archetype-sdk-tests.yml:64]: nit: Should we have new lines at end of each file?
image

**Pull request merged**